test/foundry/src/concrete/vault/OffchainAssetReceiptVault.redeem.t.sol

```diff
@@ -0,0 +1,539 @@
+        // Assume that shares is not 0
+        shares = bound(shares, 1, type(uint64).max);
+
+        OffchainAssetReceiptVault vault = createVault(alice, assetName, assetSymbol);
+        // Prank as Alice to grant role
+        vm.startPrank(alice);
+
+        vault.grantRole(vault.WITHDRAWER(), bob);
+
+        // Prank as Bob for transaction
+        vm.startPrank(bob);
+
+        uint256 expectedAssets = shares.fixedPointDiv(minShareRatio, Math.Rounding.Down);
+
+        // Get assets
+        uint256 assets = vault.previewRedeem(shares, minShareRatio);
+
+        assertEq(assets, expectedAssets);
+        // Stop the prank
+        vm.stopPrank();
+    }
+
+    /// Test withdraw function reverts without WITHDRAWER role
+    function testRedeemRevertsWithoutRole(
+        uint256 fuzzedKeyAlice,
+        uint256 shares,
+        uint256 minShareRatio,
+        bytes memory data,
+        string memory assetName,
+        string memory assetSymbol
+    ) external {
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        address alice = vm.addr((fuzzedKeyAlice % (SECP256K1_ORDER - 1)) + 1);
+        minShareRatio = bound(minShareRatio, 1, 1e18);
+        // Assume that shares is not 0
+        shares = bound(shares, 1, type(uint64).max);
+
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+
+        OffchainAssetReceiptVault vault = createVault(alice, assetName, assetSymbol);
+        vault.grantRole(vault.DEPOSITOR(), alice);
+
+        // Call the deposit function
+        vault.deposit(shares, alice, minShareRatio, data);
+
+        checkNoBalanceChange(vault, alice, alice, minShareRatio, shares, data, bytes(""));
+
+        // Stop the prank
+        vm.stopPrank();
+    }
+
+    /// Test Redeem function emits WithdrawWithReceipt event
+    function testRedeem(
+        uint256 fuzzedKeyAlice,
+        uint256 fuzzedKeyBob,
+        uint256 assets,
+        uint256 minShareRatio,
+        bytes memory data,
+        string memory assetName,
+        string memory assetSymbol
+    ) external {
+        // Ensure the fuzzed key is within the valid range for secp256k1
+        address alice = vm.addr((fuzzedKeyAlice % (SECP256K1_ORDER - 1)) + 1);
+        address bob = vm.addr((fuzzedKeyBob % (SECP256K1_ORDER - 1)) + 1);
+
+        minShareRatio = bound(minShareRatio, 0, 1e18);
+        // Assume that assets is not 0
+        assets = bound(assets, 1, type(uint256).max);
+
+        OffchainAssetReceiptVault vault = createVault(alice, assetName, assetSymbol);
+        // Prank as Alice to grant roles
+        vm.startPrank(alice);
+
+        vault.grantRole(vault.DEPOSITOR(), bob);
+        vault.grantRole(vault.WITHDRAWER(), bob);
+
+        // Prank Bob for the transaction
+        vm.startPrank(bob);
+
+        // Call the deposit function
+        vault.deposit(assets, bob, minShareRatio, data);
+
+        checkBalanceChange(vault, bob, bob, 1, assets, data);
```

Comment:
do you have any tests for redeeming some but not all of the assets deposited?

what about over several different IDs?